export/otelsdk/handler.go

```diff
@@ -1,55 +1,90 @@
+
+	"go.opentelemetry.io/auto/export"
 )
 
-// Controller handles OpenTelemetry telemetry generation for events.
-type Controller struct {
+// NewHandler returns a new configured [export.Handler] that uses the
+// OpenTelemetry SDK (go.opentelemetry.io/otel/sdk) to process and export
+// telemetry generated by auto-instrumentation.
+func NewHandler(ctx context.Context, options ...Option) (*export.Handler, error) {
+	th, err := NewTraceHandler(ctx, options...)
+	if err != nil {
+		return nil, err
+	}
+	return &export.Handler{TraceHandler: th}, nil
+}
+
+// TraceHandler handles telemetry produced by auto-instrumentation by processing
+// that telemetry with the default OpenTelemetry Go SDK.
+type TraceHandler struct {
 	logger         *slog.Logger
-	tracerProvider trace.TracerProvider
+	tracerProvider *sdk.TracerProvider
+
+	stopped atomic.Bool
 }
 
-// Trace creates a trace span for event.
-//
-// This method is safe to call concurrently.
-func (c *Controller) Trace(ss ptrace.ScopeSpans) {
+var _ export.TraceHandler = (*TraceHandler)(nil)
+
+// NewTraceHandler returns a new configured TraceHandler that uses the
+// OpenTelemetry SDK (go.opentelemetry.io/otel/sdk) to process and export
+// trace telemetry generated by auto-instrumentation.
+func NewTraceHandler(ctx context.Context, options ...Option) (*TraceHandler, error) {
+	c, err := newConfig(ctx, options)
+	if err != nil {
+		return nil, err
+	}
+
+	tp, err := c.TracerProvider(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &TraceHandler{logger: c.Logger(), tracerProvider: tp}, nil
+}
+
+// Handles the passed telemetry using the default OpenTelemetry Go SDK.
+func (h *TraceHandler) HandleTrace(scope pcommon.InstrumentationScope, url string, spans ptrace.SpanSlice) {
 	var (
 		startOpts []trace.SpanStartOption
 		eventOpts []trace.EventOption
 		endOpts   []trace.SpanEndOption
-		kvs       []attribute.KeyValue
+		spanKVs   []attribute.KeyValue
 	)
 
-	to := []trace.TracerOption{
-		trace.WithInstrumentationVersion(ss.Scope().Version()),
-		trace.WithSchemaURL(ss.SchemaUrl()),
-	}
-
-	if m := ss.Scope().Attributes(); m.Len() > 0 {
-		to = append(to, trace.WithInstrumentationAttributes(attrs(m)...))
-	}
+	tracer := h.tracerProvider.Tracer(
+		scope.Name(),
+		trace.WithInstrumentationVersion(scope.Version()),
+		trace.WithInstrumentationAttributes(attrs(scope.Attributes())...),
+		trace.WithSchemaURL(url),
+	)
 
-	tracer := c.tracerProvider.Tracer(ss.Scope().Name(), to...)
-	for k := 0; k < ss.Spans().Len(); k++ {
-		pSpan := ss.Spans().At(k)
+	for k := range spans.Len() {
+		pSpan := spans.At(k)
 
 		if pSpan.TraceID().IsEmpty() || pSpan.SpanID().IsEmpty() {
-			c.logger.Debug("dropping invalid span", "name", pSpan.Name())
+			h.logger.Debug("dropping invalid span", "name", pSpan.Name())
 			continue
 		}
-		c.logger.Debug("handling span", "tracer", tracer, "span", pSpan)
+		h.logger.Debug("handling span", "span", pSpan)
 
 		ctx := context.Background()
```

Comment:
do we want to consider `ctx` as an argument to `HandleTrace()`? (and other Handle funcs)

Reply:
Please see https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/1859#discussion_r1971098301